File: rti/internal/transport/grpc/mom.go
// MomService gRPC handler — translates rti.v1.MomService RPCs into
// calls on rti/internal/mom.Manager (via the core.ManagementObjectModel
// interface).
//
// Owner: Agent A — M12 W3 (cut-3 gRPC exposure of cut-2 mom.Manager).
//
// Composition: server.go wires a *momService into the composed Server
// via newMomService(core.ManagementObjectModel). The proto + manager
// were both frozen at M11; this handler is the wire bridge.
//
// Read-only contract: every RPC is a snapshot accessor over the
// manager's state. No RPC mutates MOM state. MOM mutations come from
// the runtime composition hooks (federation lifecycle hooks +
// per-federate counter increments off the dispatcher fan-out), not
// from federate calls.

package grpc

import (
	"context"

	"github.com/cbchoi/gorti/rti/internal/core"
	rtiv1 "github.com/cbchoi/gorti/rti/internal/genproto/rti/v1"
)

// momClassFederation / momClassFederate mirror mom.ClassHLAfederation /
// mom.ClassHLAfederate without importing the mom package — the handler
// binds to core.ManagementObjectModel so a non-default MOM impl wired
// at the composition root flows through unchanged. Constants kept in
// sync with rti/internal/mom/manager.go.
const (
	momClassFederation = "HLAobjectRoot.HLAmanager.HLAfederation"
	momClassFederate   = "HLAobjectRoot.HLAmanager.HLAfederate"
)

// momService is the concrete MomServiceServer impl. It embeds
// UnimplementedMomServiceServer for forward compatibility (gRPC v1.65+
// requirement; protoc-gen-go-grpc emits a deprecation warning otherwise).
//
// Phase 1 of the research-platform refactor (docs/research-platform.md
// §5.3): the handler binds to core.ManagementObjectModel instead of
// the concrete *mom.Manager so alternative implementations can be
// wired at the composition root.
type momService struct {
	rtiv1.UnimplementedMomServiceServer
	mgr core.ManagementObjectModel
}

func newMomService(mgr core.ManagementObjectModel) *momService {
	return &momService{mgr: mgr}
}

// QueryFederationAttributes returns a snapshot of the HLAfederation
// MOM object's attributes for the named federation. When the
// federation is not currently tracked, all repeated fields are
// returned empty and the federation_name is echoed back from the
// request — this matches the §10 standard's "federation does not
// exist" semantics (zero-valued attributes, not an error).
func (s *momService) QueryFederationAttributes(
	_ context.Context,
	req *rtiv1.QueryFederationAttributesRequest,
) (*rtiv1.QueryFederationAttributesResponse, error) {
	if req == nil {
		return nil, nilRequest("QueryFederationAttributes")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	attrs, ok := s.mgr.QueryFederation(core.FederationName(req.GetFederationName()))
	if !ok {
		return &rtiv1.QueryFederationAttributesResponse{
			FederationName: req.GetFederationName(),
		}, nil
	}
	resp := &rtiv1.QueryFederationAttributesResponse{
		FederationName: string(attrs.Name),
	}
	if len(attrs.FederateHandles) > 0 {
		resp.FederateHandles = make([]uint64, len(attrs.FederateHandles))
		for i, h := range attrs.FederateHandles {
			resp.FederateHandles[i] = uint64(h)
		}
	}
	if len(attrs.FOMModuleNames) > 0 {
		resp.FomModuleNames = make([]string, len(attrs.FOMModuleNames))
		copy(resp.FomModuleNames, attrs.FOMModuleNames)
	}
	return resp, nil
}

// QueryFederateAttributes returns one HLAfederate MOM snapshot. The
// response.found field discriminates between "federate is tracked"
// (true; remaining fields populated) and "federate not tracked"
// (false; remaining fields zero-valued). Returning an explicit found
// bool rather than a NotFound status keeps the polling pattern cheap:
// federates can poll without burning per-call status-decode work.
func (s *momService) QueryFederateAttributes(
	_ context.Context,
	req *rtiv1.QueryFederateAttributesRequest,
) (*rtiv1.QueryFederateAttributesResponse, error) {
	if req == nil {
		return nil, nilRequest("QueryFederateAttributes")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	attrs, ok := s.mgr.QueryFederate(
		core.FederationName(req.GetFederationName()),
		core.FederateHandle(req.GetFederateHandle()),
	)
	if !ok {
		return &rtiv1.QueryFederateAttributesResponse{Found: false}, nil
	}
	return &rtiv1.QueryFederateAttributesResponse{
		Found:                true,
		FederateHandle:       uint64(attrs.Handle),
		FederateName:         attrs.Name,
		FederateType:         attrs.Type,
		TimeRegulating:       attrs.TimeRegulating,
		TimeConstrained:      attrs.TimeConstrained,
		LogicalTime:          &rtiv1.LogicalTime{Value: float64(attrs.LogicalTime)},
		Lookahead:            &rtiv1.LogicalTime{Value: float64(attrs.Lookahead)},
		InteractionsSent:     attrs.InteractionsSent,
		InteractionsReceived: attrs.InteractionsReceived,
		UpdatesSent:          attrs.UpdatesSent,
		ReflectionsReceived:  attrs.ReflectionsReceived,
	}, nil
}

// EnumerateMomInstances returns the federation MOM object handle (one
// HLAfederation per federation) plus every per-federate HLAfederate
// instance. The federate ordering follows the mgr's sorted handle
// list so polling clients see a deterministic ordering across calls
// (the federation manager keeps handles dense + monotonic; the MOM
// snapshot keeps the federate-handle list sorted).
//
// Empty federation (no joined federates) returns a single MomInstance
// for the federation. Unknown federation returns an empty instances
// list — same "zero-valued" semantics as QueryFederationAttributes.
func (s *momService) EnumerateMomInstances(
	_ context.Context,
	req *rtiv1.EnumerateMomInstancesRequest,
) (*rtiv1.EnumerateMomInstancesResponse, error) {
	if req == nil {
		return nil, nilRequest("EnumerateMomInstances")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	attrs, ok := s.mgr.QueryFederation(core.FederationName(req.GetFederationName()))
	if !ok {
		return &rtiv1.EnumerateMomInstancesResponse{}, nil
	}
	resp := &rtiv1.EnumerateMomInstancesResponse{
		Instances: make([]*rtiv1.MomInstance, 0, 1+len(attrs.FederateHandles)),
	}
	resp.Instances = append(resp.Instances, &rtiv1.MomInstance{
		ClassName:    momClassFederation,
		InstanceName: string(attrs.Name),
		// FederateHandle = 0 — federation singleton has no federate handle.
	})
	for _, h := range attrs.FederateHandles {
		fed, ok := s.mgr.QueryFederate(core.FederationName(req.GetFederationName()), h)
		if !ok {
			// Federate listed in the federation snapshot but its
			// HLAfederate record was not found (race with a concurrent
			// resign). Skip — the federation snapshot is the
			// authoritative roster, and any straggler will fall off
			// on the next call.
			continue
		}
		resp.Instances = append(resp.Instances, &rtiv1.MomInstance{
			ClassName:      momClassFederate,
			FederateHandle: uint64(fed.Handle),
			InstanceName:   fed.Name,
		})
	}
	return resp, nil
}

File: pysdk/rti1516e/mom.py
"""Layer-1 SDK client for the MomService (M12 W3).

Thin async wrapper around the generated ``MomServiceStub`` that
exposes the cut-3 introspection surface for the IEEE 1516.1-2010 §10
Management Object Model: federates query the HLAfederation +
HLAfederate object snapshots tracked by the runtime.

Composition::

    async with RtiConnection.connect("grpc://...") as rti:
        async with rti.join_federation(spec, federate_name="alice") as fed:
            fed_attrs = await fed.mom.query_federation_attributes()
            print(fed_attrs.federate_handles)

            self_attrs = await fed.mom.query_federate_attributes(fed.handle)
            print(self_attrs.interactions_sent)

            instances = await fed.mom.enumerate_mom_instances()
            for inst in instances:
                print(inst.class_name, inst.federate_handle, inst.instance_name)

The ``fed.mom`` accessor (see :class:`Federate.mom` in
``connection.py``) lazily constructs one :class:`MomClient` per
federate, bound to the same gRPC channel + federation_name the
federate already holds.

Read-only contract: the proto ``MomService`` exposes only Query /
Enumerate RPCs. MOM mutations come from the runtime (federation
lifecycle hooks + per-federate counter increments off the dispatcher
fan-out), not from federate calls. See ``rti/internal/mom/manager.go``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from rti1516e._grpc_errors import translate_rpc_error

if TYPE_CHECKING:  # pragma: no cover - type-check imports only
    import grpc


@dataclass(frozen=True)
class FederationAttributes:
    """Wire-shape snapshot of one HLAfederation MOM object's attributes.

    Mirrors ``rti.v1.QueryFederationAttributesResponse``. Frozen so the
    instance is safe to hold across event-loop iterations.
    """

    federation_name: str
    federate_handles: tuple[int, ...] = ()
    fom_module_names: tuple[str, ...] = ()


@dataclass(frozen=True)
class FederateAttributes:
    """Wire-shape snapshot of one HLAfederate MOM object's attributes.

    Mirrors ``rti.v1.QueryFederateAttributesResponse``. The ``found``
    flag distinguishes "tracked + populated" from "no record (resigned
    or never joined)" without raising — matches the §10 contract that
    a query against a non-extant instance returns empty attributes.
    """

    found: bool
    federate_handle: int = 0
    federate_name: str = ""
    federate_type: str = ""
    time_regulating: bool = False
    time_constrained: bool = False
    logical_time: float = 0.0
    lookahead: float = 0.0
    interactions_sent: int = 0
    interactions_received: int = 0
    updates_sent: int = 0
    reflections_received: int = 0


@dataclass(frozen=True)
class MomInstance:
    """One MOM object instance: HLAfederation singleton or one HLAfederate.

    Mirrors ``rti.v1.MomInstance``.
    """

    class_name: str
    federate_handle: int
    instance_name: str


# IEEE 1516.1-2010 §10 standard MOM class names. Mirrors the constants
# used by the Go-side mom.Manager / mom service handler.
CLASS_HLA_FEDERATION = "HLAobjectRoot.HLAmanager.HLAfederation"
CLASS_HLA_FEDERATE = "HLAobjectRoot.HLAmanager.HLAfederate"


class MomClient:
    """Federate-bound client for the MomService gRPC surface.

    Construct via :meth:`Federate.mom` (in connection.py). The
    constructor takes a live ``grpc.aio.Channel`` plus the
    federation_name the federate has already joined under; every RPC
    populates ``wire_version=WIRE_VERSION_V1`` automatically.
    """

    def __init__(
        self,
        channel: grpc.aio.Channel,
        *,
        federation_name: str,
        federate_handle: int,
    ) -> None:
        # Lazy import — the generated stubs live under
        # rti1516e/_generated/ and are only on sys.path after the
        # GrpcTransport's ``_ensure_generated_path`` ran.
        from rti.v1 import mom_pb2_grpc

        self._stub = mom_pb2_grpc.MomServiceStub(channel)
        self._federation_name = federation_name
        self._federate_handle = int(federate_handle)

    async def query_federation_attributes(self) -> FederationAttributes:
        """Return the HLAfederation MOM object snapshot (§10.2).

        For an unknown federation the runtime returns zero-valued
        attributes (federation_name echoed from the request, empty
        lists). The MomClient surfaces this as a FederationAttributes
        with empty tuples; callers may distinguish via the empty
        federate_handles list when they want a "does this federation
        exist" probe.
        """
        from rti.v1 import common_pb2, mom_pb2

        req = mom_pb2.QueryFederationAttributesRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
        )
        try:
            resp = await self._stub.QueryFederationAttributes(req)
        except Exception as exc:  # noqa: BLE001 — translate_rpc_error reraises
            translate_rpc_error(exc)
        return FederationAttributes(
            federation_name=str(resp.federation_name),
            federate_handles=tuple(int(h) for h in resp.federate_handles),
            fom_module_names=tuple(str(n) for n in resp.fom_module_names),
        )

    async def query_federate_attributes(
        self, federate_handle: int
    ) -> FederateAttributes:
        """Return one HLAfederate MOM object snapshot (§10.3).

        ``federate_handle`` may be any handle the runtime currently
        tracks (including the caller's own ``fed.handle``). When the
        runtime does not track the requested handle, the response's
        ``found`` flag is False and remaining fields are zero-valued.
        """
        from rti.v1 import common_pb2, mom_pb2

        req = mom_pb2.QueryFederateAttributesRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            federate_handle=int(federate_handle),
        )
        try:
            resp = await self._stub.QueryFederateAttributes(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
        if not resp.found:
            return FederateAttributes(found=False)
        # logical_time / lookahead are nested LogicalTime messages; the
        # generated proto returns the nested message even when the
        # field is absent on the wire (proto3 default). Reading .value
        # on the nested object returns 0.0 in that case.
        return FederateAttributes(
            found=True,
            federate_handle=int(resp.federate_handle),
            federate_name=str(resp.federate_name),
            federate_type=str(resp.federate_type),
            time_regulating=bool(resp.time_regulating),
            time_constrained=bool(resp.time_constrained),
            logical_time=float(resp.logical_time.value),
            lookahead=float(resp.lookahead.value),
            interactions_sent=int(resp.interactions_sent),
            interactions_received=int(resp.interactions_received),
            updates_sent=int(resp.updates_sent),
            reflections_received=int(resp.reflections_received),
        )

    async def enumerate_mom_instances(self) -> list[MomInstance]:
        """Return the list of MOM object instances for the federation.

        Always begins with one HLAfederation entry (the singleton) when
        the federation is tracked, followed by one HLAfederate per
        joined federate. Ordering is deterministic — federates appear
        in sorted-handle order — so polling clients see stable rows
        across calls.
        """
        from rti.v1 import common_pb2, mom_pb2

        req = mom_pb2.EnumerateMomInstancesRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
        )
        try:
            resp = await self._stub.EnumerateMomInstances(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
        return [
            MomInstance(
                class_name=str(inst.class_name),
                federate_handle=int(inst.federate_handle),
                instance_name=str(inst.instance_name),
            )
            for inst in resp.instances
        ]
